source/common/ssl/context_manager_impl.h

```diff
@@ -22,25 +22,19 @@ class ContextManagerImpl final : public ContextManager {
   ContextManagerImpl(Runtime::Loader& runtime) : runtime_(runtime) {}
   ~ContextManagerImpl();
 
-  /**
-   * Allocated contexts are owned by the caller. However, we need to be able to iterate them for
-   * admin purposes. When a caller frees a context it will tell us to release it also from the list
-   * of contexts.
-   */
-  void releaseContext(Context* context);
-
   // Ssl::ContextManager
-  Ssl::ClientContextPtr createSslClientContext(Stats::Scope& scope,
-                                               const ClientContextConfig& config) override;
-  Ssl::ServerContextPtr
+  Ssl::ClientContextSharedPtr createSslClientContext(Stats::Scope& scope,
+                                                     const ClientContextConfig& config) override;
+  Ssl::ServerContextSharedPtr
   createSslServerContext(Stats::Scope& scope, const ServerContextConfig& config,
                          const std::vector<std::string>& server_names) override;
   size_t daysUntilFirstCertExpires() const override;
   void iterateContexts(std::function<void(const Context&)> callback) override;
 
 private:
+  void removeEmptyContexts();
   Runtime::Loader& runtime_;
-  std::list<Context*> contexts_;
+  std::list<std::weak_ptr<Context>> contexts_;
   mutable std::shared_timed_mutex contexts_lock_;
```

Comment:
Is this lock needed anymore? I'm pretty sure all operations happen on the main thread now...

Reply:
I don't think this PR changed the thread models calling this, no? To minimize risks, I'll defer removing this mutex to #3700.

Reply:
I think it did change the thread model. Previously, the lock was only required because IIRC the context was indirectly owned by ClusterInfo, which can be released on any thread. With this change I'm pretty sure that all of the interactions with the context list/map only happen on the main thread, thus the lock is no longer necessary (which is a really nice improvement IMO). If that's the case, I would remove the lock as part of this change as IMO it's a logical part of this change and we can deploy/test it discretely.

Reply:
Ah you're right, now when the context is released we're not immediately removing it from `contexts_` but wait until next create, which happens in main thread. So we do not need a lock for `contexts_` any more. @qiwzhang can you remove this mutex and lock?

Reply:
Done